src/lib/storage/frame_of_reference_column.cpp

```diff
@@ -0,0 +1,88 @@
+                                                     std::shared_ptr<const BaseCompressedVector> offset_values,
+                                                     std::shared_ptr<const pmr_vector<bool>> null_values)
+    : _block_minima{block_minima}, _offset_values{offset_values}, _null_values{null_values} {}
+
+template <typename T, typename U>
+std::shared_ptr<const pmr_vector<T>> FrameOfReferenceColumn<T, U>::block_minima() const {
+  return _block_minima;
+}
+
+template <typename T, typename U>
+std::shared_ptr<const BaseCompressedVector> FrameOfReferenceColumn<T, U>::offset_values() const {
+  return _offset_values;
+}
+
+template <typename T, typename U>
+std::shared_ptr<const pmr_vector<bool>> FrameOfReferenceColumn<T, U>::null_values() const {
+  return _null_values;
+}
+
+template <typename T, typename U>
+const AllTypeVariant FrameOfReferenceColumn<T, U>::operator[](const ChunkOffset chunk_offset) const {
+  PerformanceWarning("operator[] used");
+
+  DebugAssert(chunk_offset != INVALID_CHUNK_OFFSET, "Passed chunk offset must be valid.");
+
+  if ((*_null_values)[chunk_offset]) {
+    return NULL_VALUE;
+  }
+
+  const auto minimum = (*_block_minima)[chunk_offset / block_size];
+
+  auto decoder = _offset_values->create_base_decoder();
+  const auto value = static_cast<T>(decoder->get(chunk_offset)) + minimum;
+
+  return value;
+}
+
+template <typename T, typename U>
+size_t FrameOfReferenceColumn<T, U>::size() const {
+  return _offset_values->size();
+}
+
+template <typename T, typename U>
+std::shared_ptr<BaseColumn> FrameOfReferenceColumn<T, U>::copy_using_allocator(
+    const PolymorphicAllocator<size_t>& alloc) const {
+  auto new_block_minima = pmr_vector<T>{*_block_minima, alloc};
+  auto new_offset_values = _offset_values->copy_using_allocator(alloc);
+  auto new_null_values = pmr_vector<bool>{*_null_values, alloc};
+
+  auto new_block_minima_ptr = std::allocate_shared<pmr_vector<T>>(alloc, std::move(new_block_minima));
+  auto new_null_values_ptr = std::allocate_shared<pmr_vector<bool>>(alloc, std::move(new_null_values));
+  return std::allocate_shared<FrameOfReferenceColumn>(alloc, new_block_minima_ptr, new_offset_values,
+                                                      new_null_values_ptr);
+}
+
+template <typename T, typename U>
+size_t FrameOfReferenceColumn<T, U>::estimate_memory_usage() const {
+  static const auto bits_per_byte = 8u;
+
+  return sizeof(*this) + sizeof(typename decltype(_block_minima)::element_type) + _offset_values->data_size() +
+         _null_values->size() / bits_per_byte;
+}
+
+template <typename T, typename U>
+EncodingType FrameOfReferenceColumn<T, U>::encoding_type() const {
+  return EncodingType::FrameOfReference;
+}
+
+template <typename T, typename U>
+CompressedVectorType FrameOfReferenceColumn<T, U>::compressed_vector_type() const {
+  return _offset_values->type();
+}
+
+template class FrameOfReferenceColumn<int32_t>;
+template class FrameOfReferenceColumn<int64_t>;
```

Comment:
This encoding looks like it is more sensible to signed vs unsigned. Would this still work if we added `uint32_t`/`uint_64_t`?

Reply:
As far as I can see it would still work with the same limitation for `uint64_t` as there is for `int64_t`.